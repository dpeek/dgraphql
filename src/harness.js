import fs from 'fs'
import path from 'path'

import { graphql } from 'graphql'

import { buildSchema } from './schema'

import type { ClientConfig } from './client'

export async function init (config: ClientConfig) {
  const schemaPath = path.resolve(__dirname, '__tests__/schema.graphql')
  const sourcePath = path.resolve(__dirname, '__tests__/source.graphql')

  const source = fs.readFileSync(sourcePath).toString()

  const time = String(new Date().getTime() - 1495660000000)
  let commonVariables = { time }

  // can't use the same schema as some tests are in relay mode, the query is not
  const dataSchema = buildSchema(fs.readFileSync(schemaPath).toString(), {})
  const data = await graphql({
    schema: dataSchema,
    source,
    variableValues: commonVariables,
    contextValue: { language: 'en' }
  })
  const walk = node => {
    if (node.id && node.name) {
      commonVariables[node.name.split(' ')[0].toLowerCase()] = node.id
    }
    Object.values(node).forEach(value => {
      if (typeof value === 'object') walk(value)
    })
  }
  walk(data)

  const schema = buildSchema(fs.readFileSync(schemaPath).toString(), config)
  return (source, variables, language) => {
    return graphql({
      schema,
      source,
      variableValues: { ...commonVariables, ...(variables || {}) },
      contextValue: { language: language || 'en' }
    })
  }
}
