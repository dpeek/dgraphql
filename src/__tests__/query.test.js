import { init } from './harness'

var sequence
var vars

const create = `mutation Test($time: Int) {
  ben: createPerson(input: {
    name: "Ben Jim",
    employed: true,
    age: 37,
    height: 1.7,
    time: $time
  }) {
    person {
      id
      name
    }
  }
  ant: createPerson(input: {
    name: "Ant Tim",
    employed: true,
    age: 36,
    height: 1.76,
    time: $time
  }) {
    person {
      id
      name
    }
  }
  cat: createPerson(input: {
    name: "Cat Jim",
    employed: false,
    age: 38,
    height: 1.63,
    time: $time
  }) {
    person {
      id
      name
    }
  }
}`

beforeAll(async () => {
  const test = await init()
  sequence = test.sequence
  const time = Math.round(Math.random() * 1000000000)
  vars = await sequence([create], { time: time })
})

test('queries node field', () => {
  const query1 = `mutation Test {
    createPerson(input: { name: "David" }) {
      person {
        id
        name
      }
    }
  }`
  const query2 = `query Test ($david: ID!) {
    person(id: $david) {
      name
    }
  }`
  return sequence([query1, query2])
})

test('queries node fields', () => {
  const query1 = `mutation Test {
    createPerson(input: { name: "David", employed: true }) {
      person {
        id
        name
      }
    }
  }`
  const query2 = `query Test ($david: ID!) {
    person(id: $david) {
      name
      employed
    }
  }`
  return sequence([query1, query2])
})

test('returns null for node', () => {
  const query = `query Test {
    person(id: "0x00") {
      name
    }
  }`
  return sequence([query])
})

test('returns null for non-existing edge', () => {
  const query1 = `mutation Test {
    createPerson(input: { name: "Olivia" }) {
      person {
        id
        name
      }
    }
  }`
  const query2 = `query Test ($olivia: ID!) {
    person(id: $olivia) {
      name
      partner {
        name
      }
    }
  }`
  return sequence([query1, query2])
})

test('returns empty array for non-existing edges', () => {
  const query1 = `mutation Test {
    createPerson(input: { name: "Olivia" }) {
      person {
        id
        name
      }
    }
  }`
  const query2 = `query Test ($olivia: ID!) {
    person(id: $olivia) {
      name
      children {
        name
      }
    }
  }`
  return sequence([query1, query2])
})

test('queries type name', () => {
  const query1 = `mutation Test {
    createPerson(input: { name: "David" }) {
      person {
        id
        name
      }
    }
  }`
  const query2 = `query Test ($david: ID!) {
    person(id: $david) {
      name
      __typename
    }
  }`
  return sequence([query1, query2])
})

test('queries aliased field', () => {
  const query1 = `mutation Test {
    createPerson(input: { name: "David" }) {
      person {
        id
        name
      }
    }
  }`
  const query2 = `query Test ($david: ID!) {
    person(id: $david) {
      alias: name
    }
  }`
  return sequence([query1, query2])
})

test('queries nested field', () => {
  const query1 = `mutation Test {
    createPerson(input: { name: "David", partner: { name: "Amy" }}) {
      person {
        id
        name
      }
    }
  }`
  const query2 = `query Test ($david: ID!) {
    person(id: $david) {
      name
      partner {
        name
      }
    }
  }`
  return sequence([query1, query2])
})

test('order edges by string ascending', () => {
  const query = `query Test ($time: Int) {
    people(
      order: name_asc,
      filter: {
        time_eq: $time
      }
    ) {
      name
    }
  }`
  return sequence([query], vars)
})

test('order edges by string descending', () => {
  const query = `query Test ($time: Int) {
    people(
      order: name_desc,
      filter: {
        time_eq: $time
      }
    ) {
      name
    }
  }`
  return sequence([query], vars)
})

test('filters by string with all terms', () => {
  const query = `query Test ($time: Int) {
    people(
      order: name_desc,
      filter: {
        time_eq: $time,
        name_allofterms: "Ben Jim"
      }
    ) {
      name
    }
  }`
  return sequence([query], vars)
})

test('filters by string with any of terms', () => {
  const query = `query Test ($time: Int) {
    people(
      order: name_desc,
      filter: {
        time_eq: $time,
        name_anyofterms: "Ben Cat"
      }
    ) {
      name
    }
  }`
  return sequence([query], vars)
})

test('filters by string equal to', () => {
  const query = `query Test ($time: Int) {
    people(
      order: name_desc,
      filter: {
        time_eq: $time,
        name_eq: "Ben Jim"
      }
    ) {
      name
    }
  }`
  return sequence([query], vars)
})

test('filters by boolean equal to', () => {
  const query = `query Test ($time: Int) {
    people(
      order: name_desc,
      filter: {
        time_eq: $time,
        employed_eq: true
      }
    ) {
      name
    }
  }`
  return sequence([query], vars)
})

test('order edges by int ascending', () => {
  const query = `query Test ($time: Int) {
    people(
      order: age_asc,
      filter: {
        time_eq: $time
      }
    ) {
      name
      age
    }
  }`
  return sequence([query], vars)
})

test('order edges by int descending', () => {
  const query = `query Test ($time: Int) {
    people(
      order: age_desc,
      filter: {
        time_eq: $time
      }
    ) {
      name
      age
    }
  }`
  return sequence([query], vars)
})

test('filters by int equal to', () => {
  const query = `query Test ($time: Int) {
    people(
      order: name_desc,
      filter: {
        time_eq: $time,
        age_eq: 37
      }
    ) {
      name
      age
    }
  }`
  return sequence([query], vars)
})

test('filters by int less than', () => {
  const query = `query Test ($time: Int) {
    people(
      order: name_desc,
      filter: {
        time_eq: $time,
        age_lt: 37
      }
    ) {
      name
      age
    }
  }`
  return sequence([query], vars)
})

test('filters by int less than or equal to', () => {
  const query = `query Test ($time: Int) {
    people(
      order: name_desc,
      filter: {
        time_eq: $time,
        age_le: 37
      }
    ) {
      name
      age
    }
  }`
  return sequence([query], vars)
})

test('filters by int greater than', () => {
  const query = `query Test ($time: Int) {
    people(
      order: name_desc,
      filter: {
        time_eq: $time,
        age_gt: 37
      }
    ) {
      name
      age
    }
  }`
  return sequence([query], vars)
})

test('filters by int greater than or equal to', () => {
  const query = `query Test ($time: Int) {
    people(
      order: name_desc,
      filter: {
        time_eq: $time,
        age_ge: 37
      }
    ) {
      name
      age
    }
  }`
  return sequence([query], vars)
})

test('filters by float equal to', () => {
  const query = `query Test ($time: Int) {
      people(
        order: height_asc,
        filter: {
          time_eq: $time,
          height_eq: 1.70
        }
      ) {
        name
        height
      }
    }`
  return sequence([query], vars)
})

test('filters by float less than', () => {
  const query = `query Test ($time: Int) {
      people(
        order: height_asc,
        filter: {
          time_eq: $time,
          height_lt: 1.70
        }
      ) {
        name
        height
      }
    }`
  return sequence([query], vars)
})

test('order edges by float ascending', () => {
  const query = `query Test ($time: Int) {
    people(
      order: height_asc,
      filter: {
        time_eq: $time
      }
    ) {
      name
      height
    }
  }`
  return sequence([query], vars)
})

test('order edges by float descending', () => {
  const query = `query Test ($time: Int) {
    people(
      order: height_desc,
      filter: {
        time_eq: $time
      }
    ) {
      name
      height
    }
  }`
  return sequence([query], vars)
})

test('filters by float less than or equal to', () => {
  const query = `query Test ($time: Int) {
      people(
        order: height_asc,
        filter: {
          time_eq: $time,
          height_le: 1.70
        }
      ) {
        name
        height
      }
    }`
  return sequence([query], vars)
})

test('filters by float greater than', () => {
  const query = `query Test ($time: Int) {
      people(
        order: height_asc,
        filter: {
          time_eq: $time,
          height_gt: 1.70
        }
      ) {
        name
        height
      }
    }`
  return sequence([query], vars)
})

test('filters by float greater than or equal to', () => {
  const query = `query Test ($time: Int) {
      people(
        order: height_asc,
        filter: {
          time_eq: $time,
          height_ge: 1.70
        }
      ) {
        name
        height
      }
    }`
  return sequence([query], vars)
})
