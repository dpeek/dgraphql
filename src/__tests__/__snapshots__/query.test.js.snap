// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
Object {
  "data": Object {
    "ant": Object {
      "person": Object {
        "name": "Ant Tim",
      },
    },
    "ben": Object {
      "person": Object {
        "name": "Ben Jim",
      },
    },
    "cat": Object {
      "person": Object {
        "name": "Cat Jim",
      },
    },
  },
}
`;

exports[`filters by boolean equal to 1`] = `
Object {
  "data": Object {
    "people": Object {
      "0": Object {
        "name": "Ben Jim",
      },
      "1": Object {
        "name": "Ant Tim",
      },
    },
  },
}
`;

exports[`filters by float equal to 1`] = `
Object {
  "data": Object {
    "people": Object {
      "0": Object {
        "height": 1.7,
        "name": "Ben Jim",
      },
    },
  },
}
`;

exports[`filters by float greater than 1`] = `
Object {
  "data": Object {
    "people": Object {
      "0": Object {
        "height": 1.76,
        "name": "Ant Tim",
      },
    },
  },
}
`;

exports[`filters by float greater than or equal to 1`] = `
Object {
  "data": Object {
    "people": Object {
      "0": Object {
        "height": 1.7,
        "name": "Ben Jim",
      },
      "1": Object {
        "height": 1.76,
        "name": "Ant Tim",
      },
    },
  },
}
`;

exports[`filters by float less than 1`] = `
Object {
  "data": Object {
    "people": Object {
      "0": Object {
        "height": 1.63,
        "name": "Cat Jim",
      },
    },
  },
}
`;

exports[`filters by float less than or equal to 1`] = `
Object {
  "data": Object {
    "people": Object {
      "0": Object {
        "height": 1.63,
        "name": "Cat Jim",
      },
      "1": Object {
        "height": 1.7,
        "name": "Ben Jim",
      },
    },
  },
}
`;

exports[`filters by int equal to 1`] = `
Object {
  "data": Object {
    "people": Object {
      "0": Object {
        "age": 37,
        "name": "Ben Jim",
      },
    },
  },
}
`;

exports[`filters by int greater than 1`] = `
Object {
  "data": Object {
    "people": Object {
      "0": Object {
        "age": 38,
        "name": "Cat Jim",
      },
      "1": Object {
        "age": 37,
        "name": "Ben Jim",
      },
    },
  },
}
`;

exports[`filters by int greater than or equal to 1`] = `
Object {
  "data": Object {
    "people": Object {
      "0": Object {
        "age": 38,
        "name": "Cat Jim",
      },
      "1": Object {
        "age": 37,
        "name": "Ben Jim",
      },
    },
  },
}
`;

exports[`filters by int less than 1`] = `
Object {
  "data": Object {
    "people": Object {
      "0": Object {
        "age": 37,
        "name": "Ben Jim",
      },
      "1": Object {
        "age": 36,
        "name": "Ant Tim",
      },
    },
  },
}
`;

exports[`filters by int less than or equal to 1`] = `
Object {
  "data": Object {
    "people": Object {
      "0": Object {
        "age": 37,
        "name": "Ben Jim",
      },
      "1": Object {
        "age": 36,
        "name": "Ant Tim",
      },
    },
  },
}
`;

exports[`filters by string equal to 1`] = `
Object {
  "data": Object {
    "people": Object {
      "0": Object {
        "name": "Ben Jim",
      },
    },
  },
}
`;

exports[`filters by string with all terms 1`] = `
Object {
  "data": Object {
    "people": Object {
      "0": Object {
        "name": "Ben Jim",
      },
    },
  },
}
`;

exports[`filters by string with any of terms 1`] = `
Object {
  "data": Object {
    "people": Object {
      "0": Object {
        "name": "Cat Jim",
      },
      "1": Object {
        "name": "Ben Jim",
      },
    },
  },
}
`;

exports[`order edges by float ascending 1`] = `
Object {
  "data": Object {
    "people": Object {
      "0": Object {
        "height": 1.63,
        "name": "Cat Jim",
      },
      "1": Object {
        "height": 1.7,
        "name": "Ben Jim",
      },
      "2": Object {
        "height": 1.76,
        "name": "Ant Tim",
      },
    },
  },
}
`;

exports[`order edges by float descending 1`] = `
Object {
  "data": Object {
    "people": Object {
      "0": Object {
        "height": 1.76,
        "name": "Ant Tim",
      },
      "1": Object {
        "height": 1.7,
        "name": "Ben Jim",
      },
      "2": Object {
        "height": 1.63,
        "name": "Cat Jim",
      },
    },
  },
}
`;

exports[`order edges by int ascending 1`] = `
Object {
  "data": Object {
    "people": Object {
      "0": Object {
        "age": 36,
        "name": "Ant Tim",
      },
      "1": Object {
        "age": 37,
        "name": "Ben Jim",
      },
      "2": Object {
        "age": 38,
        "name": "Cat Jim",
      },
    },
  },
}
`;

exports[`order edges by int descending 1`] = `
Object {
  "data": Object {
    "people": Object {
      "0": Object {
        "age": 38,
        "name": "Cat Jim",
      },
      "1": Object {
        "age": 37,
        "name": "Ben Jim",
      },
      "2": Object {
        "age": 36,
        "name": "Ant Tim",
      },
    },
  },
}
`;

exports[`order edges by string ascending 1`] = `
Object {
  "data": Object {
    "people": Object {
      "0": Object {
        "name": "Ant Tim",
      },
      "1": Object {
        "name": "Ben Jim",
      },
      "2": Object {
        "name": "Cat Jim",
      },
    },
  },
}
`;

exports[`order edges by string descending 1`] = `
Object {
  "data": Object {
    "people": Object {
      "0": Object {
        "name": "Cat Jim",
      },
      "1": Object {
        "name": "Ben Jim",
      },
      "2": Object {
        "name": "Ant Tim",
      },
    },
  },
}
`;

exports[`queries aliased field 1`] = `
Object {
  "data": Object {
    "createPerson": Object {
      "person": Object {
        "name": "David",
      },
    },
  },
}
`;

exports[`queries aliased field 2`] = `
Object {
  "data": Object {
    "person": Object {
      "alias": "David",
    },
  },
}
`;

exports[`queries nested field 1`] = `
Object {
  "data": Object {
    "createPerson": Object {
      "person": Object {
        "name": "David",
      },
    },
  },
}
`;

exports[`queries nested field 2`] = `
Object {
  "data": Object {
    "person": Object {
      "name": "David",
      "partner": Object {
        "name": "Amy",
      },
    },
  },
}
`;

exports[`queries node field 1`] = `
Object {
  "data": Object {
    "createPerson": Object {
      "person": Object {
        "name": "David",
      },
    },
  },
}
`;

exports[`queries node field 2`] = `
Object {
  "data": Object {
    "person": Object {
      "name": "David",
    },
  },
}
`;

exports[`queries node fields 1`] = `
Object {
  "data": Object {
    "createPerson": Object {
      "person": Object {
        "name": "David",
      },
    },
  },
}
`;

exports[`queries node fields 2`] = `
Object {
  "data": Object {
    "person": Object {
      "employed": true,
      "name": "David",
    },
  },
}
`;

exports[`queries type name 1`] = `
Object {
  "data": Object {
    "createPerson": Object {
      "person": Object {
        "name": "David",
      },
    },
  },
}
`;

exports[`queries type name 2`] = `
Object {
  "data": Object {
    "person": Object {
      "__typename": "Person",
      "name": "David",
    },
  },
}
`;

exports[`returns empty array for non-existing edges 1`] = `
Object {
  "data": Object {
    "createPerson": Object {
      "person": Object {
        "name": "Olivia",
      },
    },
  },
}
`;

exports[`returns empty array for non-existing edges 2`] = `
Object {
  "data": Object {
    "person": Object {
      "children": Object {},
      "name": "Olivia",
    },
  },
}
`;

exports[`returns null for node 1`] = `
Object {
  "data": Object {
    "person": null,
  },
  "errors": Object {
    "0": Object {
      "locations": Object {
        "0": Object {
          "column": 5,
          "line": 2,
        },
      },
      "message": "There is no 'Node' with id '0x00'",
      "path": Object {
        "0": "person",
      },
    },
  },
}
`;

exports[`returns null for non-existing edge 1`] = `
Object {
  "data": Object {
    "createPerson": Object {
      "person": Object {
        "name": "Olivia",
      },
    },
  },
}
`;

exports[`returns null for non-existing edge 2`] = `
Object {
  "data": Object {
    "person": Object {
      "name": "Olivia",
      "partner": null,
    },
  },
}
`;
